br/pkg/task/stream.go

```diff
@@ -941,38 +944,81 @@ func RunStreamTruncate(c context.Context, g glue.Glue, cmdName string, cfg *Stre
+	p := console.StartProgressBar(
+		"Clearing Data Files", len(removed),
+		glue.WithTimeCost(),
 		glue.WithConstExtraField("kv-count", kvCount),
 		glue.WithConstExtraField("kv-size", fmt.Sprintf("%d(%s)", totalSize, units.HumanSize(float64(totalSize)))),
 	)
+	defer p.Close()
 	worker := utils.NewWorkerPool(128, "delete files")
-	wg := new(sync.WaitGroup)
+	eg, cx := errgroup.WithContext(ctx)
+	const keepFirstNFailure = 16
+	var notDeleted struct {
+		item []string
+		sync.Mutex
+	}
 	for _, f := range removed {
-		if !cfg.DryRun {
-			wg.Add(1)
-			finalFile := f
-			worker.Apply(func() {
-				defer wg.Done()
-				if err := storage.DeleteFile(ctx, finalFile.Path); err != nil {
-					log.Warn("File not deleted.", zap.String("path", finalFile.Path), logutil.ShortError(err))
-					console.Print("\n"+em(finalFile.Path), "not deleted, you may clear it manually:", warn(err))
+		worker.ApplyOnErrorGroup(eg, func() error {
+			// NOTE: We cannot get the returned value from deleting, so polling the context here again.
+			//       Maybe we'd better use an `ErrGroup` and fail fast for some types of error (or if len(notDeleted) > N)?
+			//       (It would be a disaster if we have waiting for 6hrs and eventually found out nothing deleted :| )
+			if cx.Err() != nil {
+				p.Close()
+				return cx.Err()
+			}
+			defer p.Inc()
+			if cfg.DryRun {
+				return nil
+			}
+			if err := storage.DeleteFile(ctx, f.Path); err != nil {
+				log.Warn("File not deleted.", zap.String("path", f.Path), logutil.ShortError(err))
+				notDeleted.Lock()
+				defer notDeleted.Unlock()
+				notDeleted.item = append(notDeleted.item, f.Path)
+				if len(notDeleted.item) > notDeletedBecameFatalThreshold {
+					return errors.Annotatef(berrors.ErrPiTRMalformedMetadata, "too many failure when truncating")
 				}
-			})
+			}
+			return nil
+		})
+	}
+	if err := p.Wait(ctx); err != nil {
+		return err
+	}
+	if err := eg.Wait(); err != nil {
+		return err
+	}
+
+	notDeleted.Lock()
+	if len(notDeleted.item) > 0 {
+		console.Println("Files below are not deleted due to error, you may clear it manually, check log for detail error:")
+		console.Println("- Total", em(len(notDeleted.item)), "items.")
+		if len(notDeleted.item) > keepFirstNFailure {
+			console.Println("-", em(len(notDeleted.item)-keepFirstNFailure), "items omitted.")
+			// TODO: maybe don't add them at the very first.
+			notDeleted.item = notDeleted.item[:keepFirstNFailure]
+		}
+		for _, f := range notDeleted.item {
+			console.Println(f)
 		}
 	}
-	wg.Wait()
-	clearDataFileDone()
+	notDeleted.Unlock()
 
 	// remove metadata
-	removeMetaDone := console.ShowTask("Removing metadata... ", glue.WithTimeCost())
-	if !cfg.DryRun {
-		if err := metas.DoWriteBack(ctx, storage); err != nil {
-			return err
-		}
+	pw := console.StartProgressBar("Removing Metadata", metas.PendingMeta(), glue.WithTimeCost(), glue.WithConstExtraField("metas", metas.PendingMeta()))
+	defer pw.Close()
```

Comment:
Does package `mpb` support to close the second progress bar `pw` earlier than `p`?

Reply:
pw would always be open after `p` closed. 🤔

Reply:
(Note that `mpb` would stop refreshing once it has been fulfilled.)

Reply:
get